crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -1328,11 +1332,32 @@ impl AuthorityPerEpochStore {
         Ok(txns)
     }
 
-    // Placeholder implementation
-    fn should_defer(&self, _cert: &VerifiedExecutableTransaction) -> Option<DeferralKey> {
-        // placeholder constructions to silence lints
-        let _ = DeferralKey::new_for_randomness_round(0, 0);
+    fn should_defer(
+        &self,
+        cert: &VerifiedExecutableTransaction,
+        deferred_tx_digests: &HashSet<TransactionDigest>,
+        last_randomness_round: u64,
+    ) -> Option<DeferralKey> {
+        // Defer transaction if it depends on Random object.
+        if cert
+            .shared_input_objects()
+            .any(|obj| obj.id() == SUI_RANDOMNESS_STATE_OBJECT_ID)
+        {
+            // Don't re-defer randomness-using tx.
+            if deferred_tx_digests.contains(cert.digest()) {
+                return None;
+            }
+            return Some(DeferralKey::new_for_randomness_round(
+                // Deferral by two rounds guarantees that the transaction will not depend on
+                // randomness that was revealed at the time it was sequenced.
+                last_randomness_round + 2,
+                last_randomness_round,
```

Comment:
i think this needs to be the current commit round, not the last randomness round. Otherwise you will collide with keys created during the next commit if there is no new random round written by then.